File: frontend/src/stores/project.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'

interface Project {
  id: string
  name: string
  description: string
  created_at: string
  updated_at: string
  storage_path: string
}

export const useProjectStore = defineStore('project', () => {
  const projects = ref<Project[]>([])
  const currentProject = ref<Project | null>(null)
  const isAuthenticated = ref<Record<string, boolean>>({})

  async function fetchProjects() {
    try {
      // @ts-ignore
      const result = await window.go.main.ProjectHandler.ListProjects()
      projects.value = result || []
    } catch (error) {
      console.error('Failed to fetch projects:', error)
    }
  }

  async function createProject(name: string, description: string, password: string) {
    try {
      // @ts-ignore
      const project = await window.go.main.ProjectHandler.CreateProject(name, description, password)
      projects.value.unshift(project)
      return project
    } catch (error) {
      console.error('Failed to create project:', error)
      throw error
    }
  }

  async function updateProject(id: string, name: string, description: string) {
    try {
      // @ts-ignore
      await window.go.main.ProjectHandler.UpdateProject(id, name, description)
      const index = projects.value.findIndex(p => p.id === id)
      if (index !== -1) {
        projects.value[index] = { ...projects.value[index], name, description }
      }
    } catch (error) {
      console.error('Failed to update project:', error)
      throw error
    }
  }

  async function deleteProject(id: string) {
    try {
      // @ts-ignore
      await window.go.main.ProjectHandler.DeleteProject(id)
      projects.value = projects.value.filter(p => p.id !== id)
      if (currentProject.value?.id === id) {
        currentProject.value = null
      }
    } catch (error) {
      console.error('Failed to delete project:', error)
      throw error
    }
  }

  async function verifyPassword(id: string, password: string): Promise<boolean> {
    try {
      // @ts-ignore
      const valid = await window.go.main.ProjectHandler.VerifyPassword(id, password)
      if (valid) {
        isAuthenticated.value[id] = true
      }
      return valid
    } catch (error) {
      console.error('Failed to verify password:', error)
      return false
    }
  }

  function setCurrentProject(project: Project) {
    currentProject.value = project
  }

  function isProjectAuthenticated(id: string): boolean {
    return isAuthenticated.value[id] === true
  }

  return {
    projects,
    currentProject,
    isAuthenticated,
    fetchProjects,
    createProject,
    updateProject,
    deleteProject,
    verifyPassword,
    setCurrentProject,
    isProjectAuthenticated
  }
})

File: internal/handlers/project.go
package handlers

import (
	"fmt"
	"time"

	"project-document-system/internal/crypto"
	"project-document-system/internal/database"

	"github.com/google/uuid"
)

type ProjectHandler struct{}

func NewProjectHandler() *ProjectHandler {
	return &ProjectHandler{}
}

func (h *ProjectHandler) CreateProject(name, description, password string) (*database.Project, error) {
	salt, err := crypto.GenerateSalt()
	if err != nil {
		return nil, fmt.Errorf("failed to generate salt: %w", err)
	}

	passwordHash := crypto.HashPassword(password, salt)
	projectID := uuid.New().String()

	_, err = database.DB.Exec(
		`INSERT INTO projects (id, name, description, password_hash, salt, storage_path) 
		 VALUES (?, ?, ?, ?, ?, ?)`,
		projectID, name, description, passwordHash, salt, crypto.GetStoragePath(projectID),
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create project: %w", err)
	}

	return h.GetProject(projectID)
}

func (h *ProjectHandler) GetProject(id string) (*database.Project, error) {
	project := &database.Project{}
	err := database.DB.QueryRow(
		`SELECT id, name, description, password_hash, salt, created_at, updated_at, storage_path 
		 FROM projects WHERE id = ?`, id,
	).Scan(&project.ID, &project.Name, &project.Description, &project.PasswordHash,
		&project.Salt, &project.CreatedAt, &project.UpdatedAt, &project.StoragePath)
	if err != nil {
		return nil, fmt.Errorf("failed to get project: %w", err)
	}
	return project, nil
}

func (h *ProjectHandler) ListProjects() ([]database.Project, error) {
	rows, err := database.DB.Query(
		`SELECT id, name, description, created_at, updated_at, storage_path 
		 FROM projects ORDER BY updated_at DESC`,
	)
	if err != nil {
		return nil, fmt.Errorf("failed to list projects: %w", err)
	}
	defer rows.Close()

	var projects []database.Project
	for rows.Next() {
		var p database.Project
		err := rows.Scan(&p.ID, &p.Name, &p.Description, &p.CreatedAt, &p.UpdatedAt, &p.StoragePath)
		if err != nil {
			return nil, fmt.Errorf("failed to scan project: %w", err)
		}
		projects = append(projects, p)
	}
	return projects, nil
}

func (h *ProjectHandler) UpdateProject(id, name, description string) error {
	_, err := database.DB.Exec(
		`UPDATE projects SET name = ?, description = ?, updated_at = ? WHERE id = ?`,
		name, description, time.Now(), id,
	)
	if err != nil {
		return fmt.Errorf("failed to update project: %w", err)
	}
	return nil
}

func (h *ProjectHandler) DeleteProject(id string) error {
	_, err := database.DB.Exec(`DELETE FROM projects WHERE id = ?`, id)
	if err != nil {
		return fmt.Errorf("failed to delete project: %w", err)
	}
	return nil
}

func (h *ProjectHandler) VerifyPassword(id, password string) (bool, error) {
	project, err := h.GetProject(id)
	if err != nil {
		return false, err
	}

	hash := crypto.HashPassword(password, project.Salt)
	return hash == project.PasswordHash, nil
}

func (h *ProjectHandler) ChangePassword(id, oldPassword, newPassword string) error {
	project, err := h.GetProject(id)
	if err != nil {
		return err
	}

	oldHash := crypto.HashPassword(oldPassword, project.Salt)
	if oldHash != project.PasswordHash {
		return fmt.Errorf("incorrect old password")
	}

	salt, err := crypto.GenerateSalt()
	if err != nil {
		return fmt.Errorf("failed to generate salt: %w", err)
	}

	newHash := crypto.HashPassword(newPassword, salt)
	_, err = database.DB.Exec(
		`UPDATE projects SET password_hash = ?, salt = ?, updated_at = ? WHERE id = ?`,
		newHash, salt, time.Now(), id,
	)
	if err != nil {
		return fmt.Errorf("failed to change password: %w", err)
	}
	return nil
}
